### PR TITLE
Additional min proto format testing

### PR DESCRIPTION
**Pull request by cstyan**
refactor new version flag to make it easier to pick a specific format instead of having multiple flags, plus add new formats for testing

I've added lots of TODO comments that we should audit at some point before we merge into upstream main

### CODE REVIEW

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -154,7 +154,8 @@ type flagConfig struct {
 	enableNewSDManager         bool
 	enablePerStepStats         bool
 	enableAutoGOMAXPROCS       bool
-	enableSenderRemoteWrite11  bool
+	// todo: how to use the enable feature flag properly + use the remote format enum type
```
Comment (npazosmendez) [inline cmd/prometheus/main.go:157]: I suppose that once our work is ready to be merged to `main`, we'll have converged to a single format that can be behind a single feature flag, so this won't be a problem.

(non-actionable comment)

File: web/ui/package-lock.json
Comment (npazosmendez) [inline web/ui/package-lock.json:1]: This looks unrelated

**Review by npazosmendez — APPROVED** [review]
looks good, the flag refactor will be useful as we make progress.

I'm wondering if we should discard  `min-64-fixed` now, based on [your experiments](https://github.com/prometheus/prometheus/issues/13105#issuecomment-1806593500)

File: storage/remote/queue_manager.go [hunk 2]
```diff
@@ -389,6 +389,16 @@ type WriteClient interface {
...
 }
 
+type RemoteWriteFormat int64
+
+const (
+	Base1          RemoteWriteFormat = iota // original map based format
+	Min32Optimized                          // two 32bit varint plus marshalling optimization
+	Min64Fixed                              // a single fixed64 bit value, first 32 are offset and 2nd 32 are
+	Min32Fixed
+	MinBytes // two 32bit fixed, similar to optimized but not varints + no manual marshalling optimization
```
Comment (npazosmendez) [inline storage/remote/queue_manager.go]: This is actually two varints


```suggestion
	MinBytes // two varints, similar to optimized but no manual proto marshaling optimization. varints are encoded when building the struct 
```

File: storage/remote/queue_manager.go [hunk 3]
```diff
@@ -1411,17 +1436,33 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue *queue) {
...
+				symbolTable.clear()
+			case Min32Fixed:
+				nPendingSamples, nPendingExemplars, nPendingHistograms := populateMinimizedTimeSeriesFixed32(&symbolTable, batch, pendingMin32Data, s.qm.sendExemplars, s.qm.sendNativeHistograms)
+				n := nPendingSamples + nPendingExemplars + nPendingHistograms
+				s.sendMin32Samples(ctx, pendingMin32Data[:n], symbolTable.LabelsString(), nPendingSamples, nPendingExemplars, nPendingHistograms, pBuf, &buf)
+				symbolTable.clear()
+			case MinBytes:
+				nPendingSamples, nPendingExemplars, nPendingHistograms := populateMinimizedTimeSeriesBytes(&symbolTable, batch, pendingMinBytesData, s.qm.sendExemplars, s.qm.sendNativeHistograms)
+				n := nPendingSamples + nPendingExemplars + nPendingHistograms
+				s.sendMin32Samples(ctx, pendingMin32Data[:n], symbolTable.LabelsString(), nPendingSamples, nPendingExemplars, nPendingHistograms, pBuf, &buf)
```
Comment (npazosmendez) [inline storage/remote/queue_manager.go]: This should be `sendMinBytes`

**Review by npazosmendez — CHANGES_REQUESTED** [review]

**Comment by cstyan** [comment]
Results for the latest commit
2023-11-14-195602_2517x1088_scrot
2023-11-14-195611_2515x915_scrot
2023-11-14-195619_2518x1030_scrot

**Review by npazosmendez — APPROVED** [review]
To double check I ran the benchmarks locally and, at least from a short ~20min run, I'm getting similar results.


______________________

Here's my take:
- `v11-min32-fixed`: discard. 2nd worst in network by a considerable margin, with no other gains to balance (e.g. sender/receiver CPU).
- `v11-min64-fixed`: discard. Worst in network usage, 2nd worst in sender's CPU. And while the receiver's CPU looks good, it's the stat I would trust the less from our benchmarks (as in, I don't think it is representative enough for Mimir, Thanos, etc.). And even if it was, other formats have similar or better receiver CPU. 
- `v11-min-bytes`: discard. This is basically the same as `v11-min32-optimized-varint` but with a less intuitive message definition. As I mentioned [here](https://github.com/prometheus/prometheus/issues/13105#issuecomment-1808094501), I'd rather have a custom marshaller for performance than a weird proto definition. I'm pretty sure that the CPU difference in the sender is because of [this line](https://github.com/prometheus/prometheus/blob/eec966377d6774ac6b9f4c17ad65c06bf0b7b802/prompb/custom.go#L161) in the `v11-min32-optimized-varint`  version, which may be optimizable; in any case, it's no big difference IMO.
- `v11-min-len-bytes`: discard. Similar reasoning as above, but comparing it to `v11-min-len`. The sender's CPU difference is not significant and probably not a representative measurement. Even worse, for this case there seems to be a considerable difference in receiver's CPU. Although as I said I don't think that's our more reliable metric.
- `v11-min32-optimized-varint`: probably discard. Stats are a bit worse than `v11-min-len`, but not by much. The only upside I see is the decoding is slightly easier to implement: `string[int1: int1+int2]`. 
- `v11-min-len`: keep. The only downside I see is decoding is harder and the message is less intuitive; but still this is my vote. Side note: we can even try changing the encoded symbol strings `[]byte` to a `[]string` for a nicer message; the underlying bytes should be the same (I think), and the unmarshaling can be optimized to get results similar to the current ones.

File: storage/remote/queue_manager.go [hunk 4]
```diff
@@ -1430,18 +1450,20 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue *queue) {
 		case <-timer.C:
 			batch := queue.Batch()
 			if len(batch) > 0 {
```
Comment (npazosmendez) [inline storage/remote/queue_manager.go:1453]: Looks like the `case MinLen:` is missing here.

File: web/ui/package-lock.json
Comment (npazosmendez) [inline web/ui/package-lock.json:1]: This reappeared :grin:

**Review by npazosmendez — APPROVED** [review]

File: web/web.go [hunk 5]
```diff
@@ -323,6 +324,7 @@ func New(logger log.Logger, o *Options) *Handler {
 		app = h.storage
 	}
 
+	fmt.Println("rw format for handler is: ", o.RemoteWriteFormat)
```
Comment (npazosmendez) [inline web/web.go]: Do we want to keep this or was it left from debugging?

File: scripts/remotewrite11-bench/run.sh [hunk 6]
```diff
@@ -7,7 +7,8 @@ trap "trap - SIGTERM && kill -- -$$" SIGINT SIGTERM EXIT
 declare -a INSTANCES
 # (sender,receiver) pairs to run: (sender_name; sender_flags; receiver_name; receiver_flags)
 INSTANCES+=('sender-v1;;receiver-v1;')
-INSTANCES+=('sender-v11;--enable-feature rw-1-1-sender;receiver-v11;--enable-feature rw-1-1-receiver')
+INSTANCES+=('sender-v11-min32-optimized-varint;--remote-write-format 1;receiver-v11-min32-optimized-varint;--remote-write-format 1')
+INSTANCES+=('sender-v11-min-len;--remote-write-format 5;receiver-v11-min-len;--remote-write-format 5')
```
Comment (npazosmendez) [inline scripts/remotewrite11-bench/run.sh]: This should be 2 now
```suggestion
INSTANCES+=('sender-v11-min-len;--remote-write-format 2;receiver-v11-min-len;--remote-write-format 2')
```

**Comment by cstyan** [comment]
The current `tests` failure is just 
```
example_write_adapter/server.go:1: : # github.com/prometheus/prometheus/documentation/examples/remote_storage/example_write_adapter
example_write_adapter/server.go:61:22: undefined: remote.DecodeMinimizedWriteRequest (typecheck)
```
which we can't fix until everything is merged into an upstream branch, and then we update `go.mod` in `documentation/examples/remote_storage` to be a new prom version.

I don't know what the fuzzing failure is about. Maybe it will only run/pass for PRs to merge into an actual upstream branch? :shrug:

**Comment by npazosmendez** [comment]
Nice work. I say we merge and figure out the remaining ones later, in the Prometheus' PR

**Pull request merged**